x-pack/plugins/observability_onboarding/public/components/app/custom_logs/wizard/install_elastic_agent.tsx

```diff
@@ -112,34 +126,64 @@ export function InstallElasticAgent() {
     [monitoringRole?.hasPrivileges]
   );
 
-  const { data: yamlConfig = '', status: yamlConfigStatus } = useFetcher(
-    (callApi) => {
-      if (CurrentStep === InstallElasticAgent) {
-        const options = {
-          headers: {
-            authorization: `ApiKey ${installShipperSetup?.apiKeyEncoded}`,
-          },
+  const { status: saveOnboardingStateDataStatus } = useFetcher((callApi) => {
+    const {
+      onboardingId,
+      datasetName,
+      serviceName,
+      namespace,
+      customConfigurations,
+      logFilePaths,
+    } = getState();
+    if (CurrentStep === InstallElasticAgent && onboardingId) {
+      return callApi(
+        'PUT /internal/observability_onboarding/custom_logs/{onboardingId}/save',
+        {
           params: {
-            query: { id: installShipperSetup?.id ?? '' },
+            path: { onboardingId },
+            body: {
+              state: {
+                datasetName,
+                serviceName,
+                namespace,
+                customConfigurations,
+                logFilePaths,
+              },
+            },
           },
-        };
+        }
+      );
+    }
+  }, []);
+
+  const { apiKeyEncoded, onboardingId } = installShipperSetup ?? getState();
 
+  const { data: yamlConfig = '', status: yamlConfigStatus } = useFetcher(
+    (callApi) => {
+      if (
+        CurrentStep === InstallElasticAgent &&
+        apiKeyEncoded &&
+        onboardingId
+      ) {
         return callApi(
           'GET /api/observability_onboarding/elastic_agent/config 2023-05-24',
-          installShipperSetup?.apiKeyEncoded
-            ? options
-            : { params: options.params }
+          {
+            headers: { authorization: `ApiKey ${apiKeyEncoded}` },
+            params: { query: { onboardingId } },
+          }
         );
       }
     },
-    [installShipperSetup?.apiKeyEncoded]
+    [
+      apiKeyEncoded,
+      onboardingId,
+      saveOnboardingStateDataStatus === FETCH_STATUS.SUCCESS,
+    ]
   );
 
   useEffect(() => {
-    setOnboardingId(installShipperSetup?.id ?? '');
-  }, [installShipperSetup?.id]);
-
-  const apiKeyEncoded = installShipperSetup?.apiKeyEncoded;
+    setState({ ...getState(), onboardingId, apiKeyEncoded });
```

Comment:
This could be error prone since `setState` (React's version at least) is async and the state returned by `getState` may be stale when you call `setState`.

The solution is to use the callback version `setState(prevState => newState)`